enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,205 @@
+
+The self hosted DNS functionality is composed of pods running CoreDNS-mDNS in each node and a NetworkManager dispatcher script used to configure resolv.conf to point at the node's public IP address, see [self hosted DNS design](https://github.com/openshift/installer/blob/f6bb83784a835e1f4955ef41916793c0948fd4a1/docs/design/baremetal/networking-infrastructure.md#internal-dns) for more details.
+The CHNSO will be responsible only for the deployment of the CoreDNS pods while the adjustment of the resolv.conf file will remain in the MCO, in case hosted DNS services should be disabled the CoreDNS instance should be configured as a forwarding server.
+
+#### CRD definition
+
+Below is a possible option for a CRD instance of this operator where all fields are of type Enable/Disable enum:
+
+
+```yaml
+apiVersion: clusterhostednetservices.operator.io/v1alpha1
+kind: ClusterHostedNetServicesConfig
+metadata:
+  name: clusterhostednetservicesconfig-sample
+spec:
+  dns: 
+    nodesresolution: Enable
+    apiresolution: Enable
+    appsresolution: Enable    
+  loadbalancer:
+    apiintloadbalance: Enable
+    defaultingressha: Enable
+status:
+  apiserverinternalipowner: master-0
+  ingressipowner: worker-1
+```
+
+With this sample manifest the CHNSO should:
+- Create DNS hosted service that will resolve node names, api record and .apps wildcard record.
+- Provide high availability for default ingress.
+- Load balance the requests towards api-int between master nodes.
+
+From the status section we can see that master-0 node currently owns api-int VIP and worker-1 node owns ingress VIP. Additional spec and status fields may be introduced based on requirements.
+
+#### Managing  the cluster hosted networking services
+
+All the LB, DNS, and VIP components will be enabled by default in case of fresh deployment as today, while post-deployment the admin can control the services by editing the singleton CR.
+
+TODO: detailed description of self-hosted API load balancer migration to an external load balancer.
+
+### User Stories
+
+#### As an admin I want to disable the self-hosted on-prem LB and DNS stack.
+
+#### As an admin I want to use the self-hosted on-prem LB and DNS stack also in UPI deployments.
+
+#### As an admin I want to retrieve the self-hosted stack status through K8S API.
+
+
+### Test Plan
+The operator will be tested in the OpenShift CI via e2e tests triggered using the [e2e-metal-ipi](https://github.com/openshift/release/blob/master/ci-operator/step-registry/baremetalds/e2e/baremetalds-e2e-workflow.yaml) workflow defined in the OpenShift CI steps registry and also by e2e tests for openstack, ovirt and vsphere.
+
+
+### Upgrade / Downgrade Strategy
+
+Prior to this proposal, the self-hosted stack implementation was running through the MCO.
+The implementation of the self-hostd stack using the cluster-hosted-net-services operator might lead to cases during upgrade/downgrade in which two parallel components are running the self-hosted stack. 
```

Comment:
this paragraph is no strategy, just a statement of the problem. What is the solution?

Reply:
The suggested solution is under Design section in the next line.